src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java

```diff
@@ -68,21 +56,48 @@ public TranslationMessage() {
      * </ul>
      */
     public String getMessage(final Locale locale, final String component, final String key, final Object... args) {
-    	// Use source message if the message hasn't been collected/translated
-    	String source = getMessages(Locale.forLanguageTag(ConstantsKeys.SOURCE), component, false).get(key);
-    	String collectedSourceMsg = getMessages(LocaleUtility.getSourceLocale(), component, false).get(key);
-    	if (source!=null && !source.isEmpty() && !source.equals(collectedSourceMsg)) {
-			return FormatUtils.format(source, LocaleUtility.getSourceLocale(), args);
-		}
-    	
-    	String message = FormatUtils.format(getMessages(locale, component).get(key), locale, args);
-    	if (message == null || message.isEmpty()) {
-    		throw new VIPJavaClientException(FormatUtils.format(ConstantsMsg.GET_MESSAGE_FAILED, key, component, locale));
-    	}
-    	
-    	return message;
+    	return getMessageWithArgs(locale, component, key, args);
     }
-    
+
+    /**
+     * Retrieves the localized message
+     *
+     * @param locale The locale in which the message is requested to be localized
+     * @param component The Singleton component in which the message belongs
+     * @param key The key that represents the message
+     * @param args Named arguments to replace placeholders in the message with
+     * @return One of the items in the following priority-ordered list:
+     * @throws VIPJavaClientException If none from the list below is available
+     * <ul>
+     * 		<li>The source message, if source message hasn't been collected and translated</li>
+     * 		<li>The message in the requested locale</li>
+     * 		<li>The message in the next available fallback locale</li>
+     * 		<li>The source message</li>
+     * </ul>
+     */
+    public String getMessage(final Locale locale, final String component, final String key, final Map<String, Object> args) {
+        return getMessageWithArgs(locale, component, key, args);
+    }
+
+    private String getMessageWithArgs(final Locale locale, final String component, final String key, final Object args) {
+        // Use source message if the message hasn't been collected/translated
+        String source = getMessages(Locale.forLanguageTag(ConstantsKeys.SOURCE), component, false).getMessages().get(key);
+        if (source!=null && !source.isEmpty()) {
```

Comment:
It's the same logic, just different preference, so no need to change this for now. Feel free to update it in another PR.

Reply:
I mean line 92~98 should be in this if block, too. Since when source is empty there is no need to get translation.

Reply:
It's also better surround line 84~98 with try-catch to avoid NullPointerException.

Reply:
What you said is invalid. Even if source is empty/null, it should be able to proceed with getting the translation. An example of this is in microservices that do not carry the source bundle with them. such is the case for VCF and Atlas products. When source is null or empty, just skips this if logic, and proceeds. here is no reason to fail the entire thing.

By the way, this has been like this even before this PR.

Reply:
In TranslationMessage.java, it is most likely that we do not want to hide any NullPointerExceptions because we want the product dev teams to get the exception  and fix it. Which part of the code do you think will throw a NullPointerException and when? If it is something that has to be fixed by product devs, then let the exception be thrown. Do not hide it in a catch block.

Reply:
> What you said is invalid. Even if source is empty/null, it should be able to proceed with getting the translation. An example of this is in microservices that do not carry the source bundle with them. such is the case for VCF and Atlas products. When source is null or empty, it just skips this "if" block, and proceeds. There is no reason to fail the entire thing.
> 
> By the way, this has been like this even before this PR.

With such detail case, I can understand the logic. 

By the way, though some codes exist before this PR, but I may not review them before, so I may have questions and propose them and discuss with you here. Or I don't know when these code need be changed and at then I can have chance to propose my questions.

Reply:
> In TranslationMessage.java, it is most likely that we do not want to hide any NullPointerExceptions because we want the product dev teams to get the exception and fix it. Which part of the code do you think will throw a NullPointerException and when? If it is something that has to be fixed by product devs, then let the exception be thrown. Do not hide it in a catch block.

At first sight line 84 and 86 mostly will throw NullPointerException, but after double check Service layer code, I know no such case will happen.